src/main/scala/tilelink/Nodes.scala

```diff
@@ -32,17 +32,52 @@ object TLImp extends NodeImp[TLClientPortParameters, TLManagerPortParameters, TL
     pu.copy(managers = pu.managers.map { m => m.copy (nodePath = node +: m.nodePath) })
 }
 
-case class TLClientNode(portParams: Seq[TLClientPortParameters])(implicit valName: ValName) extends SourceNode(TLImp)(portParams)
-case class TLManagerNode(portParams: Seq[TLManagerPortParameters])(implicit valName: ValName) extends SinkNode(TLImp)(portParams)
+trait TLFormatNode extends BaseNode
+{
+  def edges: Edges[List[TLEdgeIn], List[TLEdgeOut]]
+  override def formatNode() = "TileLink Node\n" + edges.out.map(currEdge =>
+  currEdge.client.clients.map(currClient => "Output Edges (edges.out.map.client.clients.map):" +
+  "\nClient Name = " + currClient.name +
+  "\nvisibility = " + currClient.visibility +
+  (if (currClient.knownToEmit == None) "\nEmits parameters are UNKNOWN\n" else
+  "\nemitsAcquireT = " + currClient.knownToEmit.get.emitsAcquireT +
+  "\nemitsAcquireB = " + currClient.knownToEmit.get.emitsAcquireB +
+  "\nemitsArithmetic = " + currClient.knownToEmit.get.emitsArithmetic +
+  "\nemitsLogical = " + currClient.knownToEmit.get.emitsLogical +
+  "\nemitsGet = " + currClient.knownToEmit.get.emitsGet +
+  "\nemitsPutFull = " + currClient.knownToEmit.get.emitsPutFull +
+  "\nemitsPutPartial = " + currClient.knownToEmit.get.emitsPutPartial +
+  "\nemitsHint = " + currClient.knownToEmit.get.emitsHint + "\n")).mkString).mkString + edges.in.map(currEdge =>
+  "Input Edge Manager Beatbytes (edges.in.map.manager.beatBytes): = " + currEdge.manager.beatBytes + "\n" + currEdge.manager.managers.map(currManager =>
+  "Input Edges (edges.in.map.manager.managers.map):" +
+  "\nManager Name = " + currManager.name +
+  "\nManager Address = " + currManager.address +
+  "\nsupportsAcquireT = " + currManager.supportsAcquireT +
+  "\nsupportsAcquireB = " + currManager.supportsAcquireB +
+  "\nsupportsArithmetic = " + currManager.supportsArithmetic +
+  "\nsupportsLogical = " + currManager.supportsLogical +
+  "\nsupportsGet = " + currManager.supportsGet +
+  "\nsupportsPutFull = " + currManager.supportsPutFull +
+  "\nsupportsPutPartial = " + currManager.supportsPutPartial +
+  "\nsupportsHint = " + currManager.supportsHint + "\n").mkString).mkString
```

Comment:
Are all these different implementations of `formatNode` actually different from each other? Can it just take some extra argument that the sub classes supply instead of copy/pasting all this code?
Pattern is something like:
```
override def formatNode() = super.formatNode("thingThatIsDifferent")
```

Reply:
Oh it's the type of `def edges: Edges[List[TLEdgeIn], List[TLEdgeOut]]` that is different? hmm

Reply:
can you maybe do a `private def formatAgentParams(client: Seq[TLClientParameters], managers: Seq[TLManagerParameters])` with all the string formatting and then the different node types only handle unpacking those lists from their unique Edge classes?